lib/Dialect/FIRRTL/Transforms/InferMemories.cpp

```diff
@@ -0,0 +1,731 @@
+  SmallVector<Operation *> opsToDelete;
+
+  /// This tracks how the result of a subfield operation which is indexes a
+  /// MemoryPortOp is used.  This is used to track if the subfield operation
+  /// needs to be cloned to access a memories rdata or wdata.
+  DenseMap<Operation *, MemDirAttr> subfieldDirs;
+
+  /// This maps a subfield-like operation from a MemoryPortOp to a new subfield
+  /// operation which can be used to read from the MemoryOp. This is used to
+  /// update any operations to read from the new memory.
+  DenseMap<Value, Value> rdataValues;
+
+  /// This maps a subfield-like operation from a MemoryPortOp to a new subfield
+  /// operation which can be used to write to the memory, the mask value which
+  /// should be set to 1, and the the correspoding wmode port of the memory
+  /// which should be set to 1.  Not all memories have wmodes, so this field
+  /// can be null. This is used to update operations to write to the new memory.
+  struct WDataInfo {
+    Value data;
+    Value mask;
+    Value mode;
+  };
+  DenseMap<Value, WDataInfo> wdataValues;
+};
+} // end anonymous namespace
+
+/// Performs the callback for each leaf element of a value.  This will create
+/// any subindex and subfield operations needed to access the leaf values of the
+/// aggregate value.
+static void forEachLeaf(ImplicitLocOpBuilder &builder, Value value,
+                        llvm::function_ref<void(Value)> func) {
+  auto type = value.getType();
+  if (auto bundleType = type.dyn_cast<BundleType>()) {
+    for (size_t i = 0, e = bundleType.getNumElements(); i < e; ++i)
+      forEachLeaf(builder, builder.create<SubfieldOp>(value, i), func);
+  } else if (auto vectorType = type.dyn_cast<FVectorType>()) {
+    for (size_t i = 0, e = vectorType.getNumElements(); i != e; ++i)
+      forEachLeaf(builder, builder.create<SubindexOp>(value, i), func);
+  } else {
+    func(value);
+  }
+}
+
+/// Drive a value to all leafs of the input aggregate value. This only makes
+/// sense when all leaf values have the same type, since the same value will be
+/// connected to each leaf. This does not work for aggregates with flip types.
+static void connectLeafsTo(ImplicitLocOpBuilder &builder, Value bundle,
+                           Value value) {
+  forEachLeaf(builder, bundle,
+              [&](Value leaf) { builder.create<ConnectOp>(leaf, value); });
+}
+
+/// Connect each leaf of an aggregate type to invalid.  This does not support
+/// aggregates with flip types.
+void InferMemoriesPass::emitInvalid(ImplicitLocOpBuilder &builder,
+                                    Value value) {
+  forEachLeaf(builder, value, [&](Value leaf) {
+    builder.create<ConnectOp>(leaf, getInvalid(leaf.getType()));
+  });
+}
+
+/// Combines two MemDirAttrs into a new one which maintains the union of the
+/// Read and Write options.
+static MemDirAttr mergeDirection(MemDirAttr a, MemDirAttr b) {
+  if (a == MemDirAttr::ReadWrite || b == MemDirAttr::ReadWrite)
```

Comment:
Would it makes sense to define `MemDirAttr` as an enum of "1, 2, 3" so we can `or` these together?

I think this helper can get hoisted up to a header or something if it is small

Reply:
Yeah, this makes sense. They were already defined as "1, 2, 3", so I just added `operation|` and `operator|=` in `FIRRTLOps.h` where we tend to put a lot of this supporting stuff.